Src/HALAL/Services/InputCapture/InputCapture.cpp

```diff
@@ -0,0 +1,92 @@
+map<Pin, IC::Instance> IC::instances_data = {
+		{PA0, IC::Instance(PA0, &htim2, TIM_CHANNEL_1, TIM_CHANNEL_2)}
+};
+map<uint8_t, IC::Instance> IC::instances = {};
+
+IC::Instance::Instance(Pin p, TIM_HandleTypeDef* tim, uint32_t ch_r, uint32_t ch_f) {
+	pin = p;
+	timer = tim;
+	channel_rising = ch_r;
+	channel_falling = ch_f;
+}
+
+optional<uint8_t> IC::register_ic(Pin& pin){
+ 	if (!IC::instances_data.contains(pin)) { return nullopt; }
+
+	Pin::register_pin(pin, ALTERNATIVE);
+	uint8_t id = IC::id_manager.front();
+	IC::id_manager.pop_front();
+
+ 	IC::Instance data = IC::instances_data[pin];
+	IC::instances[id] = data;
+	IC::instances[id].id = id;
+	return id;
+}
+
+void IC::unregister_ic(uint8_t id){
+	Pin::unregister_pin(IC::instances[id].pin);
+	IC::instances.erase(id);
+	IC::id_manager.push_front(id);
+}
+
+void IC::turn_on_ic(uint8_t id){
+	IC::Instance instance = IC::instances[id];
+	HAL_TIM_IC_Start_IT(instance.timer, instance.channel_rising);
+	HAL_TIM_IC_Start(instance.timer, instance.channel_falling);
+}
+
+void IC::turn_off_ic(uint8_t id){
+	IC::Instance instance = IC::instances[id];
+	HAL_TIM_IC_Stop_IT(instance.timer, instance.channel_rising);
+	HAL_TIM_IC_Stop(instance.timer, instance.channel_falling);
+}
+
+uint32_t IC::read_frequency(uint8_t id) {
+	IC::Instance instance = IC::instances[id];
+	return instance.frequency;
+}
+
+uint8_t IC::read_duty_cycle(uint8_t id) {
+	IC::Instance instance = IC::instances[id];
+	return instance.duty_cycle;
+}
+
+IC::Instance IC::find_instance_by_channel(uint32_t channel) {
+	for (auto id_instance: IC::instances) {
+		uint32_t& ch_rising = id_instance.second.channel_rising;
+		uint32_t& ch_falling = id_instance.second.channel_falling;
+
+		if(ch_rising == channel || ch_falling == channel) {
+			return id_instance.second;
+		}
+	}
+}
+
+void HAL_TIM_IC_CaptureCallback(TIM_HandleTypeDef *htim)
+{
+	uint32_t& active_channel = channel_dict[htim->Channel];
+	IC::Instance instance = IC::find_instance_by_channel(active_channel);
+
+	uint32_t rising_value = HAL_TIM_ReadCapturedValue(htim, instance.channel_rising);
+	if (rising_value != 0) {
+		uint32_t ref_clock = HAL_RCC_GetPCLK1Freq()*2 / (instance.timer->Init.Prescaler+1);
+		uint32_t falling_value = HAL_TIM_ReadCapturedValue(htim, instance.channel_falling);
+
+		IC::instances[instance.id].frequency = ref_clock / rising_value;
+		IC::instances[instance.id].duty_cycle = (falling_value * 100) / rising_value + MAGIC_OFFSET; // Testing has shown an error of 3 on the duty cycle calculations.
```

Comment:
Este comentario no queda nada claro. Deberías poner:
`//duty_cycle + offset due to error detected during testing.` o alguna cosa así.
Aun así me parece un poco chapuza, me parece raro que siempre te de 3 menos, ¿Estas seguro que no es un problema de calculo o aluna cosa que puedas solucionar?

Reply:
![image](https://user-images.githubusercontent.com/59616818/201892387-b04f33eb-dbc9-4974-bbf2-cb4e8998b6a1.png)
El servicio de pwm va fino, es cosa de medicion, y error de calculo no puede ser porque no es nada proporcional es un offset fijo.

Reply:
Prueba a medirlo con un osciloscopio, si en el osciloscopio da bien y en internet nadie le pone un offset es culpa de tu código que te falta algo.